File: examples/security_headers_demo.py
"""Built-in middleware + per-route header add/delete.

Uses ``examples/fusion.dev.json`` (``middleware.security``, ``request_id``, …).

Run::

    cd examples && python security_headers_demo.py
"""

from pathlib import Path

from fusion_framework import header, status
from fusion_framework.api import FusionBaseApi
from fusion_framework.app import FusionApp
from fusion_framework.middleware import default_builtin_middleware
from fusion_framework.config import get_settings, load_settings_module
from fusion_framework.middleware import add_header, delete_header
from fusion_framework.route import route

HERE = Path(__file__).resolve().parent


@route("/api/[module]", tags=["demo"])
class DemoModule(FusionBaseApi):
    def get(self):
        rid = self.state.get("request_id")
        return self.response(
            {"message": "secure defaults", "request_id": rid},
            status=status.HTTP_SUCCESS,
        )

    @delete_header("X-Powered-By", "X-Framework")
    @add_header(header.cache_control("public, max-age=60"), X_Demo="1")
    def post(self):
        return self.response({"message": "custom headers"}, status=status.HTTP_SUCCESS)

    @add_header(header.LOCATION, "/api/demo")
    def patch(self):
        return self.response({"redirect_hint": True}, status=status.HTTP_SUCCESS)


def main() -> None:
    load_settings_module("settings")
    settings = get_settings()
    settings.ensure_loaded([str(HERE)])
    app = FusionApp(settings)
    for mw in default_builtin_middleware(settings):
        app.use(mw)
    app.listen()


if __name__ == "__main__":
    main()

File: examples/security_headers_demo.cs
// Built-in middleware + AddHeader / DeleteHeader demo.
// Place fusion.dev.json beside the host app (see examples/fusion.dev.json).

using FusionFramework;

[Route("api/[module]", Tags = new[] { "demo" })]
public class DemoModule : FusionBaseApi
{
    public object Get()
    {
        State.TryGetValue("request_id", out var rid);
        return Response(new { message = "secure defaults", request_id = rid?.ToString() }, Status.HTTP_SUCCESS);
    }

    [DeleteHeader("X-Powered-By", "X-Framework")]
    [AddHeader("Cache-Control", "public, max-age=60")]
    [AddHeader("X-Demo", "1")]
    public object Post() =>
        Response(new { message = "custom headers" }, Status.HTTP_SUCCESS);
}

// Route.Register<DemoModule>();
// SettingsStore.Current.EnsureLoaded(Directory.GetCurrentDirectory());
// var settings = SettingsStore.GetSettings();
// using var app = new FusionApp(settings);
// foreach (var mw in BuiltinMiddleware.FromSettings(settings))
//     app.Use(mw);
// app.Listen();
